### ISSUE TITLE
Preview feature feedback: `referentialIntegrity`

### ISSUE THREAD

**Issue by tomhoule** [prisma/prisma#9380]
Please share your feedback about the `referentialIntegrity` preview feature released in [v3.1.1](https://github.com/prisma/prisma/releases/tag/3.1.1) in this issue.

- If you encounter a bug, please open a bug report in this repo.
- If the feature is working well for you, please share this in a comment below or leave a :+1: on this issue.

If you have any questions, don't hesitate to ask them in the [#product-feedback](https://app.slack.com/client/T0MQBS8JG/C01739JGFCM/thread/C01739JGFCM-1631633975.005200) channel in the [Prisma Slack](https://slack.prisma.io/).

**Comment by 2color** [comment]
While demoing Referential Integrity during [What's new in Prisma](https://youtu.be/nLHJCzWxS5M?t=1044) I spotted a bug with version `3.1.1`. 

With the following schema connected to PlanetScale:

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}

datasource db {
  provider          = "mysql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
  referentialIntegrity = "prisma"
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id        Int     @id @default(autoincrement())
  title     String
  content   String?
  published Boolean @default(false)
  author    User?   @relation(fields: [authorId], references: [id], onDelete: SetNull)
  authorId  Int?
}
```

- `Cascade` and `Restrict` worked as expected.
- `SetNull` didn't work ([see the video](https://youtu.be/nLHJCzWxS5M?t=1097))

**Comment by SynergyEvolved** [comment]
Hi there,
I am curious as to why there is no mention of Vitess which is a CNCF graduated project, instead Prisma only mentions the commercial version of it Planetscale.   https://www.prisma.io/docs/reference/database-reference/supported-databases#preview

**Comment by janpio** [comment]
Because no one ever asked to use Prisma with Vitess, but many people asked for PlanetScale support (which is a specially configured deployment of Vitess with specific settings activated). We _are_ partially testing on Vitess, but only with configuration that makes it PlanetScale like as well. Feel free to open an issue for explicit Vitess support and we might look into doing the work to confirm this fully working as well.

**Comment by guillaumervls** [comment]
Hi, I think I found more problems (including the fact that v3.5.0 seems to completely break referential actions when `referentialIntegrity` is set to `prisma`), detailed them in #10509

Cheers

**Comment by janpio** [comment]
There currently seems to be a problem with `referentialIntegrity = "prisma"` and `onUpdate: Cascade`: https://github.com/prisma/prisma/issues/10758

**Comment by minakawa-daiki** [comment]
I am using planetscale and the following error occurs when the number of data is large.

```shell
PrismaClientUnknownRequestError: 
Invalid `prisma.room.findMany()` invocation:


Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1153, message: "rpc error: code = ResourceExhausted desc = grpc: received message larger than max (24831606 vs. 16777216)", state: "42000" })) })
```

I assume that this is probably caused by the use of IN clauses in the part of the query that should be issued with a JOIN.
```ts
await prisma.room.findMany({
        where: { status: NextAction.COMPLETE },
        include: { matches: true },
});
```
<img width="1887" alt="image" src="https://user-images.githubusercontent.com/22878067/196007687-0a3a192a-4b79-4898-84cf-b809614e6137.png">

Currently I am using take and skip to work around this, but it would be great if you could fix this as a bug.

**Comment by Jolg42** [comment]
Hi @minakawa-daiki Could you create an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml)?
If you could provide reproduction steps it would help a lot.

Note: This is maybe unrelated to the `referentialIntegrity` feature

**Comment by minakawa-daiki** [comment]
OK.
Please give me some time as I need a large amount of data to reproduce it.

**Comment by Jolg42** [comment]
@minakawa-daiki did you get a chance to reproduce it?

**Comment by minakawa-daiki** [comment]
@Jolg42 Sorry. I have not taken the time to verify it by avoiding it in other ways.

**Comment by pfried** [comment]
I do have a problem with my schema migrating from 4.4.0 to 4.6.1 (postgres):

I have a multi tenant application and therefore I do have composite keys in almost every model


    tariff_id String?
    tariff    Tariff? @relation(fields: [tenant_id, tariff_id], references: [tenant_id, id], onDelete: SetNull)

This errors with:

```
Error parsing attribute "@relation": The `onDelete` referential action of a relation must not be set to `SetNull` when a referenced field is required.
Either choose another referential action, or make the referenced fields optional.
```

The `tenant_id` is always present in all models as it scopes the items. The `tariff_id` can be set to null in case the related `Tariff` is deleted and this is what should happen here.

**Comment by Jolg42** [comment]
@pfried do you have this line in your datasource?
```
referentialIntegrity = "prisma"
// or
relationMode = "prisma"
```
If you do not have it, then it should not be related to this issue / feature but to this https://github.com/prisma/prisma/issues/16228#issuecomment-1311559768

**Comment by pfried** [comment]
@Jolg42 No i do not have set the prisma mode but nontheless I think this is a bug and I was unable to find related issues.

Also the linked issue does not really match my case. Create a new issue then?

**Comment by pfried** [comment]
Okay give me a minute, possibly actually a bug on my side uncovered

**Comment by Jolg42** [comment]
@pfried I will mark these comments off topic for this issue, I created a separate issue based on your message here: 
https://github.com/prisma/prisma/issues/16439

**Comment by janpio** [comment]
Thanks everyone for your feedback, Relation Mode (with preview feature `referentialIntegrity` went Generally Available with 4.7.0. Read more about it in the documentation: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/relation-mode
